onnx/defs/nn/defs.cc

```diff
@@ -2829,4 +2830,483 @@ ONNX_OPERATOR_SET_SCHEMA(
+            auto* up = ctx.getInputType(3);
+            if ((up == nullptr) || (!up->has_tensor_type()))
+              return false;
+            int64_t U = up->tensor_type().elem_type();
+            builder
+                .Add(
+                    U == ONNX_NAMESPACE::TensorProto_DataType_BOOL
+                        ? "AttnBias = Where(attn_mask, AttnBiasZeros, FloatInf)"
+                        : "AttnBias = Add(attn_mask, AttnBiasZeros)");
+          } else {
+            // If is_causal set to true, the attention masking is a lower triangular matrix when the mask
+            // is a square matrix. The attention masking has the form of the upper left causal bias due to
+            // the alignment when the mask is a non-square matrix.
+            // An error is thrown if both attn_mask and is_causal are set.
+            auto* is_causal_attr = ctx.getAttribute("is_causal");
+            int64_t is_causal = (is_causal_attr != nullptr) ? is_causal_attr->i() : 0;
+            if (is_causal == 1) {
+              builder.Add("TempMask = ConstantOfShape(AttnBiasShape)", "value", mktensor(1))
+                .Add("TempMaskTri = Trilu <upper = 0> (TempMask, Zero1D)")
+                .Add("AttnBias = Where(TempMaskTri, AttnBiasZeros, FloatInf)");
+            } else {
+                builder.Add("AttnBias = Identity(AttnBiasZeros)");
+            }
+          }
+          builder.Add("AttnBiasT = Cast (AttnBias)", "to", T1);
+
+          // Group Query Attention is applied if the following are satisfied
+          // 1) q_num_heads != kv_num_heads
+          // 2) q_num_heads % kv_num_heads == 0
+          // 3) kv_num_heads == k_num_heads == v_num_heads
+          builder.Add("NGQACond1 = Equal(QNumHeads, KVNumHeads)")
+              .Add("GQACond1 = Not(NGQACond1)")
+              .Add("DivNumHeads = Div(QNumHeads, KVNumHeads)")
+              .Add("IDivNumHeads = Cast(DivNumHeads)", "to", int_type)
+              .Add("RemainderNumHeads = Mod(QNumHeads, KVNumHeads)")
+              .Add("GQACond2 = Equal(RemainderNumHeads, Zero1D)")
+              .Add("GQACond = And(GQACond1, GQACond2)")
+              .Add("InterleaveDim = Where(GQACond, IDivNumHeads, One1D)")
+              .Add("InterleaveShape = Concat <axis = 0> (One1D, InterleaveDim, One1D, One1D)")
+              .Add("KAttentionInput = Tile(PresentKey, InterleaveShape)")
+              .Add("VAttentionInput = Tile(PresentValue, InterleaveShape)");
+              //.Add("KAttentionInput = Where(GQACond, KInterleaved, PresentKey)")
+              //.Add("VAttentionInput = Where(GQACond, VInterleaved, PresentValue)");
+
+          // The following pattern is applied
+          //      Q          K          V
+          //      |          |          |
+          //      |      Transpose      |
+          //      |          |          |
+          //      ---MatMul---          |
+          //            |               |
+          //   scale---Mul              |
```

Comment:
Please see earlier comment. I believe it is preferable to scale Q and K  (by sqrt(scale) each) before multiplying them ... helps reduce overflow in the MatMul stage.

Reply:
Would we need to scale V before the second matmul as well?